pkg/pgmodel/pgx.go

```diff
@@ -643,37 +643,68 @@ func runCopyFrom(conn pgxConn, in chan copyRequest) {
 		if !ok {
 			return
 		}
-		_, err := conn.CopyFrom(
-			context.Background(),
-			pgx.Identifier{dataSchema, req.table},
-			copyColumns,
-			&req.data.batch,
-		)
-		if err != nil {
-			if pgErr, ok := err.(*pgconn.PgError); ok && strings.Contains(pgErr.Message, "insert/update/delete not permitted") {
-				/* If the error was that the table is already compressed, decompress and try again. */
-				decompressErr := decompressChunks(conn, req.data, req.table)
-				if decompressErr != nil {
-					req.data.reportResults(err)
-					pendingBuffers.Put(req.data)
-					continue
-				}
-
-				req.data.batch.ResetPosition()
-				_, err = conn.CopyFrom(
-					context.Background(),
-					pgx.Identifier{dataSchema, req.table},
-					copyColumns,
-					&req.data.batch,
-				)
-			}
-		}
+		err := copyData(conn, req)
 
 		req.data.reportResults(err)
 		pendingBuffers.Put(req.data)
 	}
 }
 
+// copy data into postgres, recovering from any errors we can
+func copyData(conn pgxConn, req copyRequest) error {
+	_, err := conn.CopyFrom(
+		context.Background(),
+		pgx.Identifier{dataSchema, req.table},
+		copyColumns,
+		&req.data.batch,
+	)
+	if err != nil {
+		err = copyFromErrorFallback(conn, req, err)
+	}
+	return err
+}
+
+// certain errors are recoverable
+func copyFromErrorFallback(conn pgxConn, req copyRequest, err error) error {
```

Comment:
that's why the `copyFromErrorFallback` calls `copyData` after decompression; so if there're duplicate keys it'll be caught and handled. I don't believe it can happen the other way since we don't insert compressed data at the moment